packages/dds/tree/src/test/feature-libraries/default-schema/schemaChecker.spec.ts

```diff
@@ -0,0 +1,722 @@
+					isNodeInSchema(mapNode_oneNumber, {
+						schema: emptySchemaCollection,
+						policy: schemaPolicy,
+					}),
+					SchemaValidationErrors.Node_MissingSchema,
+				);
+			});
+
+			it(`not in schema due to missing FieldKind entry in schemaPolicy`, () => {
+				// Schema for a map node whose fields are required and must contain a number.
+				const mapNodeSchema: TreeNodeStoredSchema = new MapNodeStoredSchema(
+					getFieldSchema(FieldKinds.required, [numberNode.type]),
+				);
+
+				// numberNode.type is not in the schema collection
+				const mapNode_oneNumber = getMapNode(
+					"myNumberMapNode",
+					new Map([[brand("prop1"), [numberNode]]]),
+				);
+
+				const schemaCollection: StoredSchemaCollection = {
+					nodeSchema: new Map([
+						[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+						[mapNode_oneNumber.type, mapNodeSchema],
+					]),
+				};
+
+				assert.equal(
+					isNodeInSchema(mapNode_oneNumber, {
+						schema: schemaCollection,
+						policy: emptySchemaPolicy,
+					}),
+					SchemaValidationErrors.Field_KindNotInSchemaPolicy,
+				);
+			});
+
+			it(`not in schema if nodes are not allowed by field`, () => {
+				const fieldSchema_requiredNumberNode = getFieldSchema(FieldKinds.required, [
+					numberNode.type,
+				]);
+				const fieldSchema_requiredStringNode = getFieldSchema(FieldKinds.required, [
+					stringNode.type,
+				]);
+				const mapNodeSchema: TreeNodeStoredSchema = new MapNodeStoredSchema(
+					fieldSchema_requiredNumberNode,
+				);
+				const mapNode = getMapNode(
+					"myNumberMapNode",
+					new Map([[brand("prop1"), [numberNode]]]),
+				);
+
+				const schemaCollection = {
+					nodeSchema: new Map([
+						[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+						[mapNode.type, mapNodeSchema],
+					]),
+				};
+				const schemaPolicy = {
+					fieldKinds: new Map([
+						[fieldSchema_requiredNumberNode.kind, FieldKinds.required],
+						[fieldSchema_requiredStringNode.kind, FieldKinds.required],
+					]),
+				};
+
+				// In schema with one number node
+				assert.equal(
+					isNodeInSchema(mapNode, { schema: schemaCollection, policy: schemaPolicy }),
+					SchemaValidationErrors.NoError,
+				);
+
+				// Not in schema after adding a string node
+				mapNode.fields.set(brand("prop2"), [stringNode]);
+				assert.equal(
+					isNodeInSchema(mapNode, { schema: schemaCollection, policy: schemaPolicy }),
+					SchemaValidationErrors.Field_NodeTypeNotAllowed,
+				);
+
+				// mapNode.fields.set(brand("prop2"), []);
+			});
+		});
+
+		describe("ObjectNodeStoredSchema", () => {
+			const numberNode = getValueNode("myNumberNode", 1);
+			const stringNode = getValueNode("myStringNode", "string");
+			const booleanNode = getValueNode("myBooleanNode", false);
+			const nullNode = getValueNode("myNullNode", null);
+			const fluidHandleNode = getValueNode("myFluidHandleNode", new TestFluidHandle());
+
+			it(`in schema (children of every type)`, () => {
```

Comment:
This is an overly large and complex test, which I don't think offers clear value.

Its not testing every type as is possible for have as many types as we want: the set of possible schema types is infinite. Instead its testing a bunch of types. I don't see how this adds any value over testing  object nodes with different field schema work (ex: 2 fields that are not the same).

Reply:
Fair, it should have been "direct children of different leaf types" or something along those lines. I think there's a bit of value in trying to be exhaustive if we have a few known cases that probably won't change frequently, to make sure that we don't have weird bugs that mishandle one of them because it is particularly different from the rest (in this case FluidHandles come to mind as something that might require edge-case-handling logic _somewhere_). I'll rename the test, if you still feel it doesn't add much I can simplify it to just have two different field schemas.

Reply:
this level if abstraction actually shouldn't know if the set of leaf types is closed or not. We used to allow custom leaf types and the policy decisions to remove that were at a higher level in the stack. Your schema checker should still handle if I were to add another leaf type (ex: some more specific string type), and should not need to depend on the default leaf domain which simple-tree assumes.

Reply:
Not super deliberately, but it seems like I ended up simplifying away the attempt-at-exhaustiveness as part of the other suggested changes 😄 . The last point is a good one as well, so this seems good now.